driver/src/main/com/mongodb/GridFSUploadStreamImpl.java

```diff
@@ -0,0 +1,137 @@
+import org.bson.BsonObjectId;
+import org.bson.BsonString;
+import org.bson.types.ObjectId;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+import static com.mongodb.assertions.Assertions.notNull;
+import static com.mongodb.util.Util.toHex;
+
+final class GridFSUploadStreamImpl extends OutputStream implements GridFSUploadStream {
+    private final MongoCollection<BsonDocument> filesCollection;
+    private final MongoCollection<BsonDocument> chunksCollection;
+    private final ObjectId fileId;
+    private final String filename;
+    private final int chunkSizeInBytes;
+    private final BsonDocument metadata;
+    private final MessageDigest md5;
+    private boolean closed;
+    private byte[] buffer;
+    private long lengthInBytes;
+    private int bufferOffset;
+    private int chunkIndex;
+
+    GridFSUploadStreamImpl(final MongoCollection<BsonDocument> filesCollection, final MongoCollection<BsonDocument> chunksCollection,
+                           final ObjectId fileId, final String filename, final int chunkSizeInBytes, final BsonDocument metadata) {
+        this.filesCollection = notNull("files collection", filesCollection);
+        this.chunksCollection = notNull("chunks collection", chunksCollection);
+        this.fileId = notNull("File Id", fileId);
+        this.filename = notNull("filename", filename);
+        this.chunkSizeInBytes = chunkSizeInBytes;
+        this.metadata = metadata;
+        try {
+            md5 = MessageDigest.getInstance("MD5");
+        } catch (NoSuchAlgorithmException e) {
+            throw new MongoGridFSException("No MD5 message digest available, cannot upload file", e);
+        }
+        chunkIndex = 0;
+        bufferOffset = 0;
+        buffer = new byte[chunkSizeInBytes];
+    }
+
+    @Override
+    public ObjectId getFileId() {
+        return fileId;
+    }
+
+    @Override
+    public OutputStream getOutputStream() {
+        return this;
+    }
+
+    @Override
+    public void write(final int b) throws IOException {
+        checkClosed();
+        buffer[bufferOffset++] = (byte) (0xFF & b);
+        lengthInBytes++;
+        if (bufferOffset == chunkSizeInBytes) {
+            writeChunks();
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        checkClosed();
+        writeChunks();
+        closed = true;
+        BsonDocument fileDocument = new BsonDocument("_id", new BsonObjectId(fileId))
+                .append("length", new BsonInt64(lengthInBytes))
+                .append("chunkSize", new BsonInt32(chunkSizeInBytes))
+                .append("uploadDate", new BsonDateTime(System.currentTimeMillis()))
+                .append("md5", new BsonString(toHex(md5.digest())))
+                .append("filename", new BsonString(filename));
+
+        if (!metadata.isEmpty()) {
+            fileDocument.append("metadata", metadata);
```

Comment:
This line isn't tested in the specification for this class, or any other unit test as far as i can see.